downstairs/src/region.rs

```diff
@@ -160,71 +171,59 @@ impl Inner {
-        }
-
-        Ok(results)
-    }
+            let (block_index, hash, nonce, tag, on_disk_hash) = row?;
 
-    /// For a given block range, return all hashes since the last flush.
-    /// `get_hashes` returns a `Vec<Vec<u64>>` of length equal to `count`. Each
-    /// `Vec<u64>` inside this parent Vec contains all hashes for a single
-    /// block, ordered so the latest hash is last.
-    pub fn get_hashes(&self, block: u64, count: u64) -> Result<Vec<Vec<u64>>> {
-        // NOTE: "ORDER BY RANDOM()" would be a good --lossy addition here
-        let stmt = "SELECT block, hash FROM integrity_hashes \
-             WHERE block BETWEEN ?1 AND ?2 \
-             ORDER BY counter ASC";
-        let mut stmt = self.metadb.prepare_cached(stmt)?;
-
-        let stmt_iter =
-            stmt.query_map(params![block, block + count - 1], |row| {
-                let block: u64 = row.get(0)?;
-                let hash: Vec<u8> = row.get(1)?;
-                Ok((block, hash))
-            })?;
+            let encryption_context = if let Some(nonce) = nonce {
+                tag.map(|tag| EncryptionContext { nonce, tag })
+            } else {
+                None
+            };
 
-        let mut results = Vec::with_capacity(count as usize);
-        for _i in 0..count {
-            results.push(Vec::new());
-        }
+            let ctx = DownstairsBlockContext {
+                block_context: BlockContext {
+                    hash: u64::from_le_bytes(hash[..].try_into()?),
+                    encryption_context,
+                },
+                block: block_index,
+                on_disk_hash: u64::from_le_bytes(on_disk_hash[..].try_into()?),
+            };
 
-        for row in stmt_iter {
-            let (row_block, hash) = row?;
-            assert_eq!(hash.len(), 8);
-            results[(row_block - block) as usize]
-                .push(u64::from_le_bytes(hash[..].try_into()?));
+            results[(ctx.block - block) as usize].push(ctx);
         }
 
         Ok(results)
     }
 
     /*
-     * Given a (block, nonce, tag), append an encryption context row.
-     *
-     * For the params, keep a list of references so that copying is
-     * minimized.
+     * Append a block context row.
      */
-    pub fn tx_set_encryption_context(
+    pub fn tx_set_block_context(
         tx: &rusqlite::Transaction,
-        encryption_context_params: &(u64, &EncryptionContext),
+        block_context: &DownstairsBlockContext,
     ) -> Result<()> {
-        let (block, encryption_context) = encryption_context_params;
-
         let stmt =
-            "INSERT INTO encryption_context (counter, block, nonce, tag) \
-             VALUES ( \
-                 (SELECT IFNULL(MAX(counter), 0) + 1 FROM encryption_context WHERE block=?1), \
-                 ?1, ?2, ?3 \
-             )";
+            "INSERT INTO block_context (block, hash, nonce, tag, on_disk_hash) \
+             VALUES (?1, ?2, ?3, ?4, ?5)";
+
```

Comment:
Why not the `SELECT IFNULL ..` stuff?  Is that part of what we doing differently here with the database commands?